core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -44,6 +45,9 @@
     private final Map<String, Setting<?>> complexMatchers = new HashMap<>();
     private final Map<String, Setting<?>> keySettings = new HashMap<>();
     private final Setting.Scope scope;
+    private static final Pattern KEY_PATTERN = Pattern.compile("^([\\w\\d_-]+[.])*[\\w\\d_-]+$");
+    private static final Pattern GROUP_KEY_PATTERN = Pattern.compile("^([\\w\\d_-]+[.])+");
```

Comment:
This should be:

    ^(?:[-\\w]+[.])+$

This assumes that (a) at least one level is required and (b) there is nothing after the group key (so it is anchored to the end of the string)